feat(audio_tools): onset detection module and test

acrilique opened the pull request:
Hey, I'm not a very good DSP student but I've actually been looking for a library that does stuff like librosa but in C, specially beat tracking. So I saw your library and dedicated some time to push the work on implementing beat tracking further. 
I'll mark this PR as draft because I'm not sure of everything in here at all. I basically tried to port what `onset.py`  does in librosa and debugged a bit so the results make some sense. I used AI assistance so I would bet I'm doing something stupid somewhere. Please take a look and tell me what you think.

8g6-new reviewed (approved):
Hey, thanks a ton for this! I’d been planning to add the Librosa-style features eventually but got tied up with other projects. Really appreciate you jumping in, you did an amazing job!

I haven’t had time to fully dive into the onset code yet, but from what I can see, and with the correlation/error metrics lining up 99+% with Librosa, it looks really solid and it should be correct since data won’t lie.

Also, no worries about the AI part, I used AI too in this project and honestly, without it the docs would probably be just one paragraph.

I’ve experimented a bit with branchless code for repeated interaction loops in my current project . For example, instead of:

```c
if (n_mels % 2 == 0) {
    onset_env[t] = (temp[n_mels / 2 - 1] + temp[n_mels / 2]) * 0.5f;
} else {
    onset_env[t] = temp[n_mels / 2];
}

// you could try this 

int mid1 = temp[n_mels / 2 - 1];
int mid2 = temp[n_mels / 2];
int is_even = (n_mels % 2 == 0);
onset_env[t] = is_even * (mid1 + mid2) * 0.5f + (1 - is_even) * mid2;
```

It can help vectorize loops better. I’ve seen good improvements in my current project and would only suggest adding it here if the performance gain is significant.

8g6-new reviewed (approved):
ok

8g6-new reviewed (approved):
# Feature Suggestion: Filterbank Generator Generalization

Hey!  I've been looking through your PR for `spectral_features.c` and I really like how you aligned the Mel filterbank with **Librosa’s Slaney-style** and continuous triangular weights. It’s a nice improvement!

My original implementation was based on common online resources that use **bin-wise triangular filters** ![example](https://i.sstatic.net/amr3C.png). These are simple, widely used, but a bit more discrete per FFT bin.

Your approach, on the other hand, is **smooth and continuous**, much like Librosa. This gives a more accurate representation of the mel scale, which helps for onset and beat detection tasks.  

So, both approaches have their merits! 

---

## Why This Could Be Useful

It might be nice to **generalize the filterbank generator** so it can support both:

- **Librosa-style (continuous)** – default for backward compatibility and general use.  
- **HTK-style or bin-wise** – for people who prefer the classic approach or need it for specific tasks like speech recognition.

We could also make the **ramp/weight calculation configurable**, so you could select triangular, bin-wise, Gaussian, cosine, etc., depending on the use case.  

This would make the generator flexible, future-proof, and still keep the smooth Librosa defaults.

---

## Suggested Configuration Structure

```c
// Ramp/weight calculation types
typedef enum {
    RAMP_BINWISE,      // Original bin-wise weight calculation (https://i.sstatic.net/amr3C.png)
    RAMP_TRIANGULAR,   // Librosa-style continuous triangular, default
    // Future variants: RAMP_GAUSSIAN, RAMP_COSINE, etc.
} ramp_shape_t;

// Mel scale variants
typedef enum {
    MEL_HTK = 0,       // Original HTK-style Mel scale
    MEL_SLANEY = 1,    // Librosa-style (Slaney) Mel scale (default)
    // Future variants: MEL_GREENWOOD, etc.
} mel_variant_t;

// Bark scale variants
typedef enum {
    BARK_STANDARD = 0,   // Default Bark variant
    // Future variants: BARK_ALTERNATE, etc.
} bark_variant_t;

// Similar enums for ERB, CHIRP, CAM, LOG10 as needed

// Filterbank configuration structure
typedef struct {
    filter_type_t scale;         // F_MEL, F_BARK, F_ERB, F_CHIRP, F_CAM, F_LOG10; default: F_MEL
    int scale_variant;           // Variant ID for the selected scale; default: MEL_VARIANT_SLANEY for F_MEL
    ramp_shape_t ramp_shape;     // Default: RAMP_TRIANGULAR
    bool include_nyquist;        // Default: true (fft_size/2 + 1 bins)
    bool normalize;              // Default: true (normalize filter areas to 1)
    float fmin;                  // Minimum frequency (Hz)
    float fmax;                  // Maximum frequency (Hz)
    size_t num_filters;          // Number of filters
    size_t sample_rate;          // Sample rate (Hz)
    size_t fft_size;             // FFT size
} filterbank_config_t;

// Macro for default initialization
#define FILTERBANK_CONFIG_DEFAULT(fmin, fmax, nfilters, sr, fft_sz) \
    (filterbank_config_t) { \
        .scale = F_MEL, \
        .scale_variant = MEL_VARIANT_SLANEY, \
        .ramp_shape = RAMP_TRIANGULAR, \
        .include_nyquist = true, \
        .normalize = true, \
        .fmin = (fmin), \
        .fmax = (fmax), \
        .num_filters = (nfilters), \
        .sample_rate = (sr), \
        .fft_size = (fft_sz) \
    }
```
- You could **customize `gen_filterbank()`** however you see fit to make the final function easy to use.  
- Default behavior should remain **Slaney Mel with continuous triangular ramp** (as you implemented) for users who just want a straightforward Mel spectrogram.  
- Advanced users could **select HTK-style or bin-wise weights** via configuration options.  
- From your changes, it’s clear you **understand the project well**, so letting you tweak or extend the generator would be great.  
- This approach keeps the API **powerful, maintainable, and user-friendly** at the same time.

acrilique commented:
Hello! I addressed your comments from yesterday and removed the "Draft" status from this PR. Performance has improved quite a bit. Please take a look and leave more comments or merge if you feel like it.

acrilique commented on src/audio_tools/spectral_features.c:
```diff
@@ -273,84 +303,137 @@ void print_melbank(const filter_bank_t *v) {
+            #pragma GCC diagnostic push
+            #pragma GCC diagnostic ignored "-Wcast-function-type"
+            hz_to_scale = (double (*)(double))hz_to_mel;
+            scale_to_hz = (double (*)(double))mel_to_hz;
+            #pragma GCC diagnostic pop
+            break;
         case F_BARK:    hz_to_scale = hz_to_bark;    scale_to_hz = bark_to_hz; break;
         case F_ERB:     hz_to_scale = hz_to_erb;     scale_to_hz = erb_to_hz; break;
         case F_CHIRP:   hz_to_scale = hz_to_chirp;   scale_to_hz = chirp_to_hz; break;
         case F_CAM:     hz_to_scale = hz_to_cam;     scale_to_hz = cam_to_hz; break;
         case F_LOG10:   hz_to_scale = hz_to_log10;   scale_to_hz = log10_to_hz; break;
         default:
-            ERROR("Unknown filterbank type enum: %d", type);
-            free(bin_edges);  // Don't forget to free!
+            ERROR("Unknown filterbank type enum: %d", config->scale);
+            free(hz_edges);
             return non_zero;
     }
```
Addressed with suggested approach 1
